spring-pulsar/src/main/java/org/springframework/pulsar/observation/PulsarMessageSenderContext.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.observation;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.pulsar.observation.PulsarMessageSenderContext.MessageHolder;
+
+import io.micrometer.observation.transport.SenderContext;
+
+/**
+ * {@link SenderContext} for Pulsar messages.
+ *
+ * @author Chris Bono
+ */
+public class PulsarMessageSenderContext extends SenderContext<MessageHolder> {
+
+	private final String beanName;
+
+	private final String destination;
+
+	PulsarMessageSenderContext(MessageHolder messageHolder, String topic, String beanName) {
+		super((carrier, key, value) -> messageHolder.property(key, value));
+		setCarrier(messageHolder);
+		this.beanName = beanName;
+		this.destination = topic;
+	}
+
+	public static PulsarMessageSenderContext newContext(String topic, String beanName) {
+		MessageHolder messageHolder = new MessageHolder();
+		PulsarMessageSenderContext senderContext = new PulsarMessageSenderContext(messageHolder, topic, beanName);
+		return senderContext;
+	}
+
+	public Map<String, String> properties() {
+		return getCarrier().properties();
+	}
+
+	/**
+	 * The name of the bean sending the message (typically a {@code PulsarTemplate}).
+	 * @return the name of the bean sending the message
+	 */
+	public String getBeanName() {
+		return this.beanName;
+	}
+
+	/**
+	 * The destination topic for the message.
+	 * @return the topic the message is being sent to
+	 */
+	public String getDestination() {
+		return this.destination;
+	}
+
+	/**
+	 * Acts as a carrier for a Pulsar message and records the propagated properties for
+	 * later access by the Pulsar message builder.
+	 */
+	public static class MessageHolder {
+
+		private Map<String, String> properties = new HashMap<>();
```

Comment:
`final`?

Also, I'm not sure that you'd to have this class instances be created outside of your API.
So, where is a `private` ctor?

Reply:
I think we need to come up w/ consistent rules on when to use / not use final. It is mixed across projects currently. I think for cases where it is an instance variable and its immutable makes sense. I am not a fan of final everywhere which ultimately clutters the code. I just came from "that place" and I don't miss it. :) 

I believe Spring Boot takes this approach as well (not final everywhere).

Reply:
As far as I remember some Java compilers do an optimization for `final` properties: https://www.baeldung.com/java-final-performance

I'm not sure what you mean, though, about Spring Boot.
See, for example, its `AbstractNestedCondition`:
```
	private final ConfigurationPhase configurationPhase;

	AbstractNestedCondition(ConfigurationPhase configurationPhase) {
		Assert.notNull(configurationPhase, "ConfigurationPhase must not be null");
		this.configurationPhase = configurationPhase;
	}
```
or `PropertyMapper`:
```
	private final PropertyMapper parent;

	private final SourceOperator sourceOperator;

	private PropertyMapper(PropertyMapper parent, SourceOperator sourceOperator) {
		this.parent = parent;
		this.sourceOperator = sourceOperator;
	}
```

Reply:
Just a quick 2c: AFAIK the criteria is we always add `final` to instance / class-level properties that are immutable, while local variables and method arguments we leave without (even if they are in fact final). This is important for enforcing immutable objects, thread-safety, etc.

For local variables I think compilers detect they don't mutate and apply any optimizations.

Reply:
@tomazfernandes that makes sense and sounds reasonable to me.